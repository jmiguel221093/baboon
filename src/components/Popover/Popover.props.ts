export interface PopoverProps {
	/** Content to display in the popover */
	children?: React.ReactNode;
	/** Whether the popover is open or not */
	open?: boolean;
	/** Element to trigger the popover */
	trigger: React.ReactElement | HTMLElement;
	/** Event to detect when popover needs to be closed */
	onClose?: () => void;
	/** Fills the popover width with the trigger width size */
	fullWidth?: boolean;
	/**
	 * Whether the popover should be displayed on the left, center or right of the
	 * @default "left"
	 */
	align?: "left" | "right" | "center";
	/** Sets the popover position as fixed */
	fixed?: boolean;
	/** Whether the section should have a border or not */
	borderLess?: boolean;
}
