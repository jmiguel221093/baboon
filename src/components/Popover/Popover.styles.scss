@import "../../styles/main.scss";

$popover-background-color: var(--color-background-surface);
$popover-shadow: var(--shadow-container);
$popover-layer: var(--z-index-popover);
$popover-spacing: var(--spacing-2);
$popover-border-radius: var(--border-radius-7);
$popover-margin: var(--spacing-3);
$popover-translate-y: pxToRem(12px);

$popover-motion-duration: var(--motion-duration-250);
$popover-motion-ease: var(--motion-ease-in-out);

.Popover {
	position: absolute;
	padding: $popover-margin 0;
	opacity: 0;
	transition: opacity $popover-motion-duration $popover-motion-ease,
		transform $popover-motion-duration $popover-motion-ease;
	transform: translateY(-$popover-translate-y);
	z-index: $popover-layer;
	&__Body {
		padding: $popover-spacing;
		background-color: $popover-background-color;
		box-shadow: $popover-shadow;
		border-radius: $popover-border-radius;
	}
	&#{&}--fixed {
		position: fixed;
	}
	&#{&}--open {
		opacity: 1;
		transform: translateY(0);
	}
}
