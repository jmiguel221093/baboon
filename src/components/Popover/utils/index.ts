import setPosition from "./setPosition";
import getAlignment from "./getAlignment";

export { setPosition, getAlignment };
