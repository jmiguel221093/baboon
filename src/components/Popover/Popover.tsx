import { Children, useCallback, useEffect, useRef, useState } from "react";
import type { ReactNode } from "react";
import { createPortal } from "react-dom";
import { cssClassNames } from "../../utils";

import type { PopoverProps } from "./Popover.props";

import "./Popover.styles.scss";

const COMPONENT_NAME = "Popover";

const Popover = ({
	children,
	trigger,
	onClose,
	open,
	fullWidth,
	align = "left",
	fixed,
}: PopoverProps) => {
	const popoverRef = useRef<HTMLDivElement>(null);
	const triggerRef = useRef<HTMLDivElement>(null);
	const [popoverStyle, setPopoverStyle] = useState<React.CSSProperties>({
		top: 0,
		left: 0,
		width: undefined,
	});

	const triggerRect = triggerRef.current?.children[0].getBoundingClientRect();

	const className = cssClassNames(COMPONENT_NAME, fixed && "fixed");

	useEffect(() => {
		const handleClickOutside = (event: MouseEvent) => {
			if (
				popoverRef.current &&
				!popoverRef.current.contains(event.target as Node) &&
				triggerRef.current &&
				event.target !== triggerRef.current.children[0] &&
				!triggerRef.current.children[0].contains(event.target as Node)
			) {
				if (onClose) {
					onClose();
				}
			}
		};
		if (open) {
			document.addEventListener("click", handleClickOutside);
		}
		return () => {
			document.removeEventListener("click", handleClickOutside);
		};
	}, [open]);

	const setAlignment = useCallback(
		(popoverWidth: number) => {
			if (!triggerRect) {
				return;
			}
			if (align === "left") {
				return triggerRect.left;
			}
			if (align === "center") {
				return (
					triggerRect.left + triggerRect.width / 2 - popoverWidth / 2
				);
			}
			if (align === "right") {
				return triggerRect.left + triggerRect.width - popoverWidth;
			}
		},
		[align, open]
	);

	useEffect(() => {
		if (open && popoverRef.current && triggerRect) {
			const popoverStyleObj = structuredClone(popoverStyle);
			const popoverWidth = popoverRef.current.offsetWidth;
			const popoverHeight = popoverRef.current.offsetHeight;
			popoverStyleObj.top = triggerRect.bottom;
			popoverStyleObj.left = setAlignment(popoverWidth);
			if (window.innerHeight < triggerRect.bottom + popoverHeight) {
				popoverStyleObj.top =
					triggerRect.top - popoverHeight + window.pageYOffset;
			}
			if (fullWidth) {
				popoverStyleObj.width = triggerRect.width;
				popoverStyleObj.left = triggerRect.left;
			}
			setPopoverStyle({ ...popoverStyleObj });
		}
	}, [popoverRef, open]);

	return (
		<>
			<div ref={triggerRef}>{trigger as ReactNode}</div>
			{open &&
				createPortal(
					<div
						className={className}
						ref={popoverRef}
						style={popoverStyle}
					>
						{Children.only(children)}
					</div>,
					document.body
				)}
		</>
	);
};

export default Popover;
