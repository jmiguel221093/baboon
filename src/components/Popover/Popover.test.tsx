import { screen, render } from "@testing-library/react";
import Popover from "./Popover";

describe("Popover", () => {
	it("should be mounted", () => {
		render(<Popover />);
		expect(screen).toBeTruthy();
	});
});
