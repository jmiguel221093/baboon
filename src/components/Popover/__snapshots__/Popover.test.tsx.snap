// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover component renders without crashing 1`] = `ShallowWrapper {}`;
